compiler/rustc_parse/src/validate_attr.rs

```diff
@@ -42,16 +46,34 @@ pub fn parse_meta<'a>(sess: &'a ParseSess, attr: &Attribute) -> PResult<'a, Meta
         path: item.path.clone(),
         kind: match &item.args {
             MacArgs::Empty => MetaItemKind::Word,
-            MacArgs::Eq(_, t) => {
-                let t = TokenTree::Token(t.clone()).into();
-                let v = parse_in(sess, t, "name value", |p| p.parse_unsuffixed_lit())?;
-                MetaItemKind::NameValue(v)
-            }
             MacArgs::Delimited(dspan, delim, t) => {
                 check_meta_bad_delim(sess, *dspan, *delim, "wrong meta list delimiters");
                 let nmis = parse_in(sess, t.clone(), "meta list", |p| p.parse_meta_seq_top())?;
                 MetaItemKind::List(nmis)
             }
+            MacArgs::Eq(_, MacArgsEqKind::Ast(expr)) => {
+                if let ast::ExprKind::Lit(lit) = &expr.kind {
+                    // Even though we already have a literal, there are some restrictions on what
+                    // literal forms are allowed. So we reparse it to detect unacceptable things
+                    // like suffixes.
```

Comment:
All these restrictions are a single `lit.kind.is_unsuffixed()` check, no need to reparse IMO.

Reply:
Ok. This required duplicating the error strings from `parser/attr.rs` here, let me know if you want me to factor them out into constants or something.